### ISSUE TITLE
Substantial memory usage in dask.order

### ISSUE THREAD

**Issue by fjetter** [dask/dask#11055]
I'm currently still working on a reproducer but I've seen dask.order taking up a substantial amount of memory 

Particularly, I've seen this set https://github.com/dask/dask/blob/00a5d9200e2cee31e19d4e9fbe88ace8ad41cad0/dask/order.py#L672 grow to several GB of memory

![image](https://github.com/dask/dask/assets/8629629/40dc8a55-4456-4e35-b947-fb3c8260a61f)

This is likely not only slow but can also crash the scheduler

**Comment by slevang** [comment]
I'm running into something similar. My example has way too much going on to share for now, but I'm seeing `_connecting_to_roots` run forever until all 200GB of my memory is consumed and it crashes.

**Comment by fjetter** [comment]
@slevang is it possible to provide a minimal reproducer for this?

**Comment by slevang** [comment]
Here's a fairly minimal example derived from my workflow. It's an SVD followed by an iterative rotation step of the initial outputs of the decomposition. This may be a silly thing to do without any intermediate computations to break up the task graph, but at least it reproduces this potential issue. You should see the following use close to 10GB while scheduling:

```python
import dask.array as da
import numpy as np


def rotate(X, n_iter=100):
    gamma = 1
    n_samples, n_modes = X.shape
    R = np.eye(n_modes)

    for i in range(n_iter):
        basis = X @ R

        basis2 = basis * basis.conj()
        basis3 = basis2 * basis
        W = np.diag(np.sum(basis2, axis=0))
        alpha = gamma / n_samples

        transformed = X.conj().T @ (basis3 - (alpha * basis @ W))
        U, svals, VT = da.linalg.svd_compressed(transformed, k=n_modes, compute=False)
        R = U @ VT

    Xrot = X @ R
    return Xrot

n_modes = 100

x = da.random.random((10000, 100), chunks=(10, -1))
u, s, v = da.linalg.svd_compressed(x, k=n_modes, compute=False)

u_rot = rotate(u).compute()
```

**Comment by fjetter** [comment]
Thanks for the reproducer. I took a brief look into this. This is not immediately actionable since I doubt anybody can actually work with this code but for documentation purposes...

- The `_connecting_to_roots` function is the culprit as already pointed out in the above screenshot
- During development, I already anticipated memory blowup and implemented some set deduplication, see https://github.com/dask/dask/blob/6a04b4ed70934b548d25a793eed90f033a690531/dask/order.py#L654-L675
- This set deduplication is helping but it is also fuzzy and is not detecting all duplicates. The duplicates are blowing up and are multiplying the memory usage (There will still be some nontrivial memory usage even if that was fixed)

Apart from this, I also found by CPU profiling that [this](https://github.com/dask/dask/blob/6a04b4ed70934b548d25a793eed90f033a690531/dask/order.py#L674) assert is eating up a lot of CPU time and we should get rid of it
